### ISSUE TITLE
package-lock.json downgrades from v2 (newly released npm@7) to v1

### ISSUE THREAD

**Issue by AviVahl** [dependabot/dependabot-core#2643]
**Package manager/ecosystem**
npm (newly released 7!)

**Manifest contents prior to update**
Irrelevant.

**Updated dependency**
Any?

**What you expected to see, versus what you actually saw**
I expected dependabot to see `"lockfileVersion": 2,` in `package-lock.json` and use `npm@7`.

**Images of the diff or a link to the PR, issue or logs**
https://github.com/AviVahl/ts-tools/pull/92

**Comment by saschanaz** [comment]
Somehow it even does not actually bump the version: https://github.com/w3c/webidl2.js/pull/500

**Comment by Kjaer** [comment]
Same happening on my private repos, better to take care of this soon, because now I manually upgrade the deps :/

**Comment by mikecousins** [comment]
Yeah this one is annoying. I've had to disable dependabot now for Javascript.

**Comment by Ignigena** [comment]
I'm observing similar but slightly different behaviour when using NPM 7 in combination with the Workspaces feature.

In this scenario, for dependencies inside of one of the workspace folders, Dependabot ignores entirely the root `package-lock.json` file. It only bumps the version number or range in the `package.json` for the relevant subdirectory.

In my case I've even specified NPM 7 under the `engines` option in the root `package.json` but Dependabot seems to ignore this entirely as indicated by prior posts.

**Comment by stefcameron** [comment]
I can confirm the reported behavior. Very annoying. When the bot runs, it downgrades my lockfile to v1. Then when I pull the latest code and `npm install`, I get a `package-lock.json` change because NPM converts the lock file back to v2. And it's not just the internal version code that changes. The entire structure of the lock file changes because of significant changes made to the v2 format with `npm@7`.

The bot should detect the lock file version and use the corresponding NPM CLI.

**Comment by stefcameron** [comment]
Even more concerning is whether NPM `<=6` is actually _recreating_ the lock file at v1 rather than _downgrading_ it. That would be very bad. Means every time the bot runs, my locks are voided and replaced with current, defeating the purpose of the lock file... I can see NPM 7 _converting_, but NPM 6 or less wouldn't know about this new format, so I wouldn't be surprised if it's an overwrite rather than a downgrade that doesn't alter any of the locked versions...

**Comment by longility** [comment]
I really want to use dependabot. Like some here, I have to disable dependabot for now. I’m waiting for this to be done before enabling. 😞

**Comment by sayantan300** [comment]
Hey everyone here, u can try a feature I guess, GitHub actions itself keeps npm V6, so I added a `postinstall` script to update npm to npm@latest (V7) and then reinstall it, so you can try that in your repos if possible...

**Comment by sayantan300** [comment]
Alternative way is to manually check for updates through this:
https://www.npmjs.com/package/npm-check-updates

**Comment by BorntraegerMarc** [comment]
https://github.com/dependabot/dependabot-core/blob/main/CHANGELOG.md#v01301-14-january-2021 states that dependabot is compatible with npm v7

However I'm using dependabot native GitHub integration and for me it does not detect npm v7 lockfiles and overwrites it...

**Comment by jurre** [comment]
> https://github.com/dependabot/dependabot-core/blob/main/CHANGELOG.md#v01301-14-january-2021 states that dependabot is compatible with npm v7

That's maybe a little confusingly worded on our side, apologies. We are working on adding npm 7 support and that release allowed us to _detect_ if we should be using npm 7, but it doesn't actually run the update against npm 7 yet. We hope to land it soon though! You can follow along here: https://github.com/dependabot/dependabot-core/pull/3011

**Comment by BorntraegerMarc** [comment]
Thank you for the clarification @jurre 

Is GitHub's native Dependabot always using the latest dependabot-core version?

It would be helpful to know, because once the PR is merged we would like to upgrade npm once GitHub supports npm v7.

**Comment by jurre** [comment]
> Is GitHub's native Dependabot always using the latest dependabot-core version?


Yeah we track it pretty closely although there is a delay between releasing a version and rolling it out into production. It's a semi-manual process so sometimes we release a version and roll it out the next day.

We'll publish an update on the [GitHub changelog](https://github.blog/changelog/) when support has landed though, and I'll make sure to ping folks here as well.

**Comment by AraHaan** [comment]
@jurre is there also perhaps a way we can get an command line version of dependabot that we can deploy ourselves that increments all updates (if any) at once within github actions and bundle all updates into a single pr using the create-pull-request action that way we do not need to worry about the one built into github whenever new releases land and then make an ``setup-dependabot`` action that any workflow can basically use with the type of updates they need to run on the project (that would accept what settings are possible in the dependabot.yml file without needing to supply that file to the action)?

**Comment by jurre** [comment]
@AraHaan yeah that's totally possible and people do this, https://github.com/dependabot/dependabot-script is a good place to start for inspiration on how to run dependabot standalone, and here is a more full fledged example of running Dependabot as an action: https://github.com/CGA1123/dependabot-lein-runner

**Comment by jurre** [comment]
Npm 7 support should be [available now](https://github.blog/changelog/2021-02-08-dependabot-npm-v7-support/), if you run into issues please do let us know so we can fix them 🎉

**Comment by AviVahl** [comment]
Awesome. Closing issue then.

I wonder, which version of npm is dependabot using? or is it using @npmcli/arborist directly?
I got a weird PR from dependabot today. That lock file was *just* generated from scratch (using npm@7.5.1; which comes with latest node@15), and I already use yargs@16.2.0: https://github.com/AviVahl/ts-tools/pull/152

Reasons I can imagine causing this:
- differences of versions (npm/arborist) generating that lock file.
- mishandling of npm's new workspaces feature.
- or just bugs with the new lock file v2 handling.

Probably a separate issue though.

**Comment by jurre** [comment]
> I wonder, which version of npm is dependabot using? or is it using @npmcli/arborist directly?

We look at the `lockfileVersion` in the package.json and then run the npm cli for most operations. There are a few things we run arborist for, but typically we won't update the lockfile directly using arborist.

We'll take a look at the PR you linked to, thanks for pointing to it! I wouldn't expect it to be related to different aroborist/npm versions, they should all result in the same lockfile, but workspaces or general bugs are also my initial suspicion.

**Comment by feelepxyz** [comment]
@AviVahl thanks for the report, it looks like this issues is related  to npm workspaces where dependabot has incorrectly tried to update `yargs` in your [nested package](https://github.com/AviVahl/ts-tools/blob/main/packages/build/package.json#L16) but installed it top-level instead.

We're working on a couple of fixes.

**Comment by perrin4869** [comment]
`npm` needs an option to update a dependency inside a single workspace (which correctly updates package-lock.json)...

**Comment by feelepxyz** [comment]
> `npm` needs an option to update a dependency inside a single workspace (which correctly updates package-lock.json)...

This is possible but not well documented, you'll have to update the requirement in the nested `package.json` then run `npm install` without any arguments in the folder with the `package-lock.json`.

Working on a fix here https://github.com/dependabot/dependabot-core/pull/3106

**Comment by AraHaan** [comment]
@feelepxyz btw letting you know that every run of the github hosted dependabot now complains about updating npm even when the project does not use npm (.net core with nuget).

**Comment by feelepxyz** [comment]
> @feelepxyz btw letting you know that every run of the github hosted dependabot now complains about updating npm even when the project does not use npm (.net core with nuget).

Oh thanks for letting us know, we'll get that fixed up 👍

**Comment by trey** [comment]
Maybe a silly question, but to enable npm v7, would you just include a `dependabot.yml` file with `package-ecosystem: "npm"` configured and that's it to get the latest version? Based on the documentation, I don't see how to specify the npm version.

**Comment by feelepxyz** [comment]
Yep! Dependabot will pick the npm version based on the lockfile version.

**Comment by trey** [comment]
@feelepxyz Thanks! Then is the `dependabot.yml` necessary at all, or will it just pay attention to the existing lockfile?

**Comment by feelepxyz** [comment]
@trey ah it depends. You'll need to check in the config file to enable Dependabot [version updates](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/about-dependabot-version-updates) but won't need to do this for [security updates](https://docs.github.com/en/code-security/supply-chain-security/managing-vulnerabilities-in-your-projects-dependencies/about-dependabot-security-updates). For version updates you just need the following to enable npm6/7:

`.github/dependabot.yml`
```yaml
version: 2
updates:
  - package-ecosystem: "npm"
    directory: "/" # directory with package.json/package-lock.json
    schedule:
      interval: "daily"
```

**Comment by trey** [comment]
@feelepxyz Thanks! I think that makes sense. But even without the `dependabot.yml` file, the security updates will honor the `lockfileVersion` found in `package-lock.json`?

**Comment by feelepxyz** [comment]
> @feelepxyz Thanks! I think that makes sense. But even without the `dependabot.yml` file, the security updates will honor the `lockfileVersion` found in `package-lock.json`?

Yes that's right 👍  we pass the same options to the update and dependabot-core figures out the version of npm to use.

**Comment by trey** [comment]
@feelepxyz Thanks again for answering my (probably silly) questions!
